src/target/riscv/riscv-013.c

```diff
@@ -1784,6 +1780,37 @@ static void deinit_target(struct target *target)
 	info->version_specific = NULL;
 }
 
+static int add_ext_trigger_to_group(struct target *target, unsigned int group,
+		riscv_ext_trigger_t external_trigger)
+{
+	uint32_t write_val = DM_DMCS2_HGSELECT;
+	assert(group <= 31);
+	assert(external_trigger.dmexttrigger < 16);
```

Comment:
(Optionally)

Detail for additional safety:

```suggestion
	assert(external_trigger.dmexttrigger < 16);
	assert(external_trigger.dmexttrigger.grouptype == HALT_GROUP || external_trigger.dmexttrigger.grouptype == RESUME_GROUP);	
```

Reply:
I think the following code is safe enough.
`write_val = set_field(write_val, DM_DMCS2_GROUPTYPE, (grouptype == HALT_GROUP) ? 0 : 1);`